File: year2/systems-design-implementation/Lab6/SwimServer/Services/Service.cs
using SwimModel.Domain;
using SwimModel.Services;
using SwimServer.Repository.Interface;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SwimServer.Services
{
    public class Service : IService
    {
        private ParticipantRepository participantRepository;
        private ProbaRepository probaRepository;
        private InscriereRepository inscriereRepository;
        private UserRepository userRepository;
        private readonly IDictionary<String, Observer> loggedUsers;

        public Service(ParticipantRepository repo1, ProbaRepository repo2, InscriereRepository repo3,
            UserRepository repo4)
        {
            participantRepository = repo1;
            probaRepository = repo2;
            inscriereRepository = repo3;
            userRepository = repo4;
            loggedUsers = new Dictionary<String, Observer>();
        }


        public void Login(User user, Observer client)
        {
            User usr = null;
            foreach (User u in userRepository.FindAll())
            {
                if (u.username == user.username && u.password == user.password)
                {
                    usr = u;
                }
            }

            if (usr != null)
                loggedUsers[user.username] = client;
            else
                throw new Exception("Authentication failed.");
        }

        public void Logout(User user)
        {
            loggedUsers.Remove(user.username);
        }

        public IEnumerable<Proba> FindAllProbe()
        {
            return probaRepository.FindAll();
        }

        public Proba FindProba(int id)
        {
            return probaRepository.FindOne(id);
        }

        public Participant AddParticipant(Participant participant)
        {
            return participantRepository.Add(participant);
        }

        public Inscriere AddInscriere(Inscriere inscriere, User user)
        {
            Inscriere i = inscriereRepository.Add(inscriere);
            foreach (String username in loggedUsers.Keys)
            {
                // if (user.username != username)
                // {
                    Task.Run(() => loggedUsers[username].Update(i));
                // }
            }

            return i;
        }

        public int NrParticipanti(Proba proba)
        {
            return inscriereRepository.NrParticipanti(proba);
        }

        public IEnumerable<Participant> GetInscrisi(Proba proba)
        {
            return inscriereRepository.GetInscrisi(proba);
        }

        public IEnumerable<Proba> GetInscrieri(Participant participant)
        {
            return inscriereRepository.GetInscrieri(participant);
        }
    }
}

File: year2/systems-design-implementation/Lab5/SwimServer/src/main/java/srv/Service.java
package srv;

import domain.Inscriere;
import domain.Participant;
import domain.Proba;
import domain.User;
import repo.intf.InscriereRepository;
import repo.intf.ParticipantRepository;
import repo.intf.ProbaRepository;
import repo.intf.UserRepository;
import services.IService;
import services.Observer;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class Service implements IService {

    private ParticipantRepository participantRepository;
    private ProbaRepository probaRepository;
    private InscriereRepository inscriereRepository;
    private UserRepository userRepository;
    private Map<String, Observer> loggedUsers;
    private final int defaultThreadsNo = 5;

    public Service(ParticipantRepository participantRepository, ProbaRepository probaRepository, InscriereRepository inscriereRepository, UserRepository userRepository) {
        this.participantRepository = participantRepository;
        this.probaRepository = probaRepository;
        this.inscriereRepository = inscriereRepository;
        this.userRepository = userRepository;
        loggedUsers = new ConcurrentHashMap<>();
    }

    public void login(User user, Observer client) throws Exception {
        User usr = null;
        for (User u : userRepository.findAll()) {
            if (u.getUsername().equals(user.getUsername()) && u.getPassword().equals(user.getPassword())) {
                usr = u;
            }
        }
        if (usr == null) {
            throw new Exception("Invalid login!");
        } else {
            loggedUsers.put(usr.getUsername(), client);
        }
    }

    public void logout(User user) {
        loggedUsers.remove(user.getUsername());
    }

    public List<Proba> findAllProbe() {
        return probaRepository.findAll();
    }

    public Participant addParticipant(Participant participant) throws Exception {
        return participantRepository.add(participant);
    }

    public synchronized Inscriere addInscriere(Inscriere inscriere) throws Exception {
        Inscriere i = inscriereRepository.add(inscriere);
        ExecutorService executor = Executors.newFixedThreadPool(defaultThreadsNo);
        for (Observer client : loggedUsers.values()) {
            if (client != null)
                executor.execute(() -> {
                    try {
                        System.out.println("Notifying client " + client);
                        client.update(i);
                    } catch (Exception e) {
                        System.err.println(e);
                    }
                });
        }
        executor.shutdown();
        return i;
    }

    public int nrParticipanti(Proba proba) {
        return inscriereRepository.nrParticipanti(proba);
    }

    public List<Participant> getInscrisi(Proba proba) {
        return inscriereRepository.getInscrisi(proba);
    }

    public List<Proba> getInscrieri(Participant participant) {
        return inscriereRepository.getInscrieri(participant);
    }
}
